### PR TITLE
Make EC widget theme reactive - Update widget url when the theme changes

### PR DESCRIPTION
**Pull request by toger5**
Signed-off-by: Timo K <toger5@hotmail.de>

## Checklist

-   [x] Tests written for new code (and old code if feasible).
-   [x] New or updated `public`/`exported` symbols have accurate [TSDoc](https://tsdoc.org/) documentation.
-   [x] Linter and other CI checks pass.
-   [x] Sign-off given on the changes (see [CONTRIBUTING.md](https://github.com/matrix-org/matrix-react-sdk/blob/develop/CONTRIBUTING.md)).

### CODE REVIEW

**Review by t3chguy — CHANGES_REQUESTED** [review]
This would cause you to be kicked out of conferences, docs, games etc, we should not be causing the widget to reload without user intervention, they may lose whatever they were typing etc.

**Comment by toger5** [comment]
depends on where in the url the widget is adding the theme.
If it is adding it to the hash it will not reload the page.

This needs to be documented so that widgets that cannot just be reloaded are aware, that the theme part of the url can change on the fly.

What do you think about only updating it if the url has the theme in the hash?
Another idea to improove that situation would be to make this a widget capability. So the default would be to not change the url onChange. Widget devs have to opt in to make url change on theme change and they can only do so if that is compatible with their widget like with EC.

**Comment by toger5** [comment]
> How about sending the theme update over postMessage?

That was the initial plan but this is simpler and does not require a new msc.
(action would need changes in: react-sdk, rust-sdk, matrix-widget-api, element-call + new spec)

I still think the action would be the most conventional solution. It just would be a much bigger work piece. Maybe worthit though. I like this workaround to support the light theme soonish. It reuses the url template system and the theme property that already exists there.

File: src/components/views/elements/AppTile.tsx [hunk 1]
```diff
@@ -457,6 +464,17 @@ export default class AppTile extends React.Component<IProps, IState> {
         });
     };
 
+    private onThemeChanged = (): void => {
+        // Regenerate widget url when the theme changes
+        // this updates the url from e.g. `theme=light` to `theme=dark`
+        // We only do this with EC widgets where the theme prop is in the hash. If the widget puts the
+        // theme template variable outside the url hash this would cause a (IFrame) page reload on every theme change.
+        if (this.props.app.type === WidgetType.CALL.preferred) this.setState({ widgetUrl: this.sgWidget?.embedUrl });
```
Comment (t3chguy) [inline src/components/views/elements/AppTile.tsx:472]: Isn't there a `matches` method on WidgetType.CALL to check both?

**Review by t3chguy — CHANGES_REQUESTED** [review]
Looks like it needs test coverage to be merged

File: test/components/views/elements/__snapshots__/AppTile-test.tsx.snap [hunk 2]
```diff
@@ -84,7 +84,9 @@ exports[`AppTile for a persistent app should render 1`] = `
     id="1"
   >
     <div
+      aria-label="widget-container"
```
Comment (t3chguy) [inline test/components/views/elements/__snapshots__/AppTile-test.tsx.snap:87]: This it not an accessible label.

File: src/components/views/elements/AppTile.tsx [hunk 3]
```diff
@@ -700,7 +719,12 @@ export default class AppTile extends React.Component<IProps, IState> {
                     const zIndexAboveOtherPersistentElements = 101;
 
                     appTileBody = (
-                        <div className="mx_AppTile_persistedWrapper">
+                        <div
+                            className="mx_AppTile_persistedWrapper"
+                            // We store the widget url to make it possible to test the value of the widgetUrl. since the iframe itself wont be here. (PersistedElement are in a different dom tree)
+                            data-widget-url={this.state.widgetUrl}
```
Comment (t3chguy) [inline src/components/views/elements/AppTile.tsx:725]: Maybe use `data-test-*` so its clear its only for tests

**Review by t3chguy — CHANGES_REQUESTED** [review]

File: src/theme.ts [hunk 4]
```diff
@@ -121,8 +121,10 @@ function clearCustomTheme(): void {
     // remove all css variables, we assume these are there because of the custom theme
     const inlineStyleProps = Object.values(document.body.style);
     for (const prop of inlineStyleProps) {
-        if (prop.startsWith("--")) {
-            document.body.style.removeProperty(prop);
+        if (typeof prop === "string") {
+            if (prop.startsWith("--")) {
+                document.body.style.removeProperty(prop);
+            }
         }
```
Comment (t3chguy) [inline src/theme.ts:128]: why nested if? `&&` is a thing

File: src/settings/watchers/ThemeWatcher.ts [hunk 5]
```diff
@@ -25,7 +26,7 @@ import { findHighContrastTheme, setTheme } from "../../theme";
 import { ActionPayload } from "../../dispatcher/payloads";
 import { SettingLevel } from "../SettingLevel";
 
-export default class ThemeWatcher {
+export default class ThemeWatcher extends EventEmitter {
```
Comment (t3chguy) [inline src/settings/watchers/ThemeWatcher.ts:29]: Please use a TypedEventEmitter instead

**Review by t3chguy — APPROVED** [review]

**Pull request merged**